modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTxFinishFuture.java

```diff
@@ -485,6 +486,12 @@ private boolean finish(boolean commit,
             req.writeVersion(tx.writeVersion() != null ? tx.writeVersion() : tx.xidVersion());
 
             try {
+                if (isNull(cctx.discovery().node(n.id()))){
+                    log.error("Unable to send message (node left topology): " + n);
+
+                    fut.onNodeLeft();
```

Comment:
Seems that it's not equivalent to the logic that was used earlier:
1. Originally as part of `GridCacheIoManager.send()`  we've called `onSend()` that contains following code:
```
        if (!onSend(msg, node.id()))
            return;
```
and 
```
        if (msg.error() != null && cctx.kernalContext().isStopping())
            return false;
```
so, originally we might successfully finish future even if there's no expected node in the map (if cluster is stopping), now  we will call fut.onNodeLeft();

Besides that, we don't call `cctx.deploy().prepare((GridCacheDeployable)msg);`. Not sure whether it's important or not.

Reply:
Similar is about tracing `org.apache.ignite.internal.managers.communication.GridIoManager#send` - we won't trace message if there's no node in the map, originally we did.

Same is about checking communication listener. Originally we threw IgniteCheckedException and while catching it within 
```
 catch (IgniteCheckedException e) {
                if (!cctx.discovery().alive(node.id()) || !cctx.discovery().pingNode(node.id()))
                    throw new ClusterTopologyCheckedException("Node left grid while sending message to: " + node.id(), e);
```
we only threw ClusterTopologyCheckedException if node wasn't alive from spi point of view (or ping failure) and not discoCache one:
```
    @Nullable public ClusterNode getAlive(UUID nodeId) {
        assert nodeId != null;

        return getSpi().getNode(nodeId); // Go directly to SPI without checking disco cache.
    }
```
otherwise we've made another attempt of sending message.

Reply:
I believe that there might be other similar issues.